### PR TITLE
Requirement builder with the new data format

### PR DESCRIPTION
**Pull request by SamChou19815**
### Summary 

Depends on #587.

This PR switches the requirement building algorithm to use the new data format proposed in #522. It only builds the first graph (the graph that keeps all the edges with constraint violations) for now, since the current data enforces that there are no constraint violations. 

### Test Plan 

Load it the first time and compare your requirement progress against staging. It should be mostly the same. There might be some discrepancies with edges from requirement with warnings to AP/IB credits, since the new implementation fixes the bug in the old system (See https://github.com/cornell-dti/course-plan/pull/553#discussion_r734788977 when I first write the migration script).

Screen Shot 2021-11-14 at 17 54 59

After that, you can play around with adding courses and requirements from both the semester view and the requirement bar, with both regular and self-check requirements. Everything should work fine.

I would like to get more eyes on this since it's a pretty significant change.

### CODE REVIEW

**Comment by hahnbeelee** [comment]
When I dragged CS 4121 for the CS Practicum requirement it fulfilled both the CS Elective and CS Practicum requirement. This doesn't happen on prod.
image

**Comment by hahnbeelee** [comment]
When I added a course from the add modal from the req bar for tech elec it got counted as tech elec and cs elec
image

image

**Comment by hahnbeelee** [comment]
Weird bug where I added CS 3300 as an external spec from the req bar, but it didn't actually get added? The course card isn't in the semester view, and it isn't being counted for actually fulfilling the requirement (0/3)
image

**Comment by hahnbeelee** [comment]
Fulfilled course cards are not showing for self-select (specifically the CS major tech elec) - notice 2/3 but no courses listed
image

**Comment by hahnbeelee** [comment]
local vs. prod.
image
image

**Review by hahnbeelee — CHANGES_REQUESTED** [review]
I see the change of AP/IB not fulfilling requirements by default.
Tested + Works as desired:
Delete a course
add a course 
add semester
clear semester
delete a semester

Weird behavior with:
- selectable requirements

File: src/global-firestore-data/semesters.ts [hunk 1]
```diff
@@ -90,7 +90,7 @@ export const addCourseToSemester = (
   year: number,
   season: FirestoreSemesterSeason,
   newCourse: FirestoreSemesterCourse,
-  requirementID?: string,
+  choiceUpdater: (choice: FirestoreCourseOptInOptOutChoices) => FirestoreCourseOptInOptOutChoices,
```
[the remark this answers is not in the document]
Comment (SamChou19815) [reply src/global-firestore-data/semesters.ts:93]: It's a function that takes in the current choice as parameter and return the updated choice.

**Comment by SamChou19815** [comment]
@hahnbeelee all the above-mentioned issues should be fixed now. How I fixed them is explain in the commit message.

**Review by hahnbeelee — CHANGES_REQUESTED** [review]
Thank you! I found something else - when I add CS 4210 the modal says it's mapping to "CS Electives" but when I added it it automatically fulfilled all the requirements it could possibly fulfill. LOL
image

**Comment by SamChou19815** [comment]
> Thank you! I found something else - when I add CS 4210 the modal says it's mapping to "CS Electives" but when I added it it automatically fulfilled all the requirements it could possibly fulfill. LOL image

Should be fixed in the last commit. Turns out a filter condition is inverted 😅

**Review by hahnbeelee — CHANGES_REQUESTED** [review]
I think this is my last round of comments.

File: src/requirements/requirement-frontend-utils.ts [hunk 2]
```diff
@@ -291,28 +298,41 @@ const computeFulfillmentStatistics = (
...
+        if (arbitraryOptInSlotNames.has(slotName)) {
           // the user wants to use this course to override this sub-requirement
           coursesThatFulfilledSubRequirements[subRequirementIndex].push(courseTaken);
           subRequirementProgress[subRequirementIndex] +=
             fulfilledBy === 'courses' ? 1 : courseTaken.credits;
           // don't break, in case the user wants to override more sub-requirements with the same course
         } else if (
           eligibleCourses[subRequirementIndex].includes(courseTaken.courseId) &&
-          subRequirementProgress[subRequirementIndex] < perSlotMinCount[subRequirementIndex] &&
-          !(optOutSlotNames && (fulfilledBy === 'credits' || optOutSlotNames.has(slotName)))
```
[the remark this answers is not in the document]
Comment (SamChou19815) [reply src/requirements/requirement-frontend-utils.ts:315]: Yep, I don't think we need this. The new requirement flow will never opt-out of a single slot, but instead the entire requirement.

File: src/requirements/admin/requirement-graph-admin-utils.ts [hunk 3]
```diff
@@ -3,7 +3,7 @@ import { checkNotNull } from '../../utilities';
 import {
   semestersCollection,
   toggleableRequirementChoicesCollection,
-  selectableRequirementChoicesCollection,
```
[the remark this answers is not in the document]
Comment (SamChou19815) [reply src/requirements/admin/requirement-graph-admin-utils.ts:6]: All migrated away

**Review by hahnbeelee — APPROVED** [review]
LGTM!

**Pull request merged**